### DISCUSSION TITLE
NaN detected for output grad_gamma_p

### DISCUSSION THREAD

**Discussion by martinbb0** [casadi/casadi#3505] [Q&A] [answered]
Hi! I'm running an Optimal Control problem with a MINLP formulation that seems to converge to a good solution. However, I'm getting a warning I don't understand. I don't know what `grad_gamma_p` refers to, and I don't see why the NaN is generated. I would like to learn how to inspect this.

> CasADi -  WARNING("solver:nlp_grad failed: NaN detected for output grad_gamma_p, at (row 6, col 1).") [.../casadi/core/oracle_function.cpp:265]
> CasADi -  WARNING("Failed to calculate multipliers") [.../casadi/core/nlpsol.cpp:621]

For context, the problematic constraints and variables are the following:
```
opti.debug.g_describe(6)
opti.debug.x_describe(1)
```

> 'Opti constraint of shape 5x1, defined at C:\Users\....
>        opti.subject_to( X(:,k+1) == F(X(:,k),[U(k,:),param(k,1),param(k,3)]) );
>      At nonzero 1.'
> 'Opti decision variable 'opti4_x_1' of shape 4x1, defined at C:\Users\...
>        Chiller_on = opti.variable(hp); % Chiller state (on/off - discrete variable)
>      At nonzero 1.'

Where `Chiller_on` is the discrete variable of the OC problem, and the constraints are the gap-closing multiple shooting constraints of the system dynamics, described by the function F.
F is defined as:
```
%% Define symbolic variables of the model
x = MX.sym('x',Tank.seg.n);
    % Tank segment temperatures.
u = MX.sym('u', 4); 
    % 1: m_chiller ; 2: Tchiller_out  (actuation variables in the MPC)
    % 3: Tamb ; 4: m_load   (parameters in the MPC)
dTdt = MX.zeros(Tank.seg.n,1);

% Auxiliary variables (dependant on x and u)
m_tank = u(1) - u(4);
up = 1./(1+exp(-nu1*m_tank));
down = 1./(1+exp(nu1*m_tank));
tbot = up.*u(2) + down.*x(1);           
tsup = up.*u(2) + down.*(u(1).*u(2) - m_tank.*tbot).*u(4)./(u(4).*u(4) + eps); 
treturn = tsup + deltaT; 
ttop = down.*treturn + up.*x(end); 
Qf_tes = G.*(x(1:end-1) - x(2:end));

%% Define continuous system: dot(x) = f(x,u)
% Temperature time derivative equations (x_dot) 
dTdt(1) = 1./mcp.*( (UA_lid + UA_wall).*(u(3) - x(1)) - 4/3*Qf_tes(1) ...
        + m_tank.*cp.*((tbot - x(1)).*up + (x(1) - x(2)).*down) );          % Bottom segment
dTdt(end) = 1./mcp.*( (UA_lid + UA_wall).*(u(3) - x(end)) + 4/3*Qf_tes(end) ...
        + m_tank.*cp.*((x(end-1) - x(end)).*up + (x(end) - ttop).*down) );  % Top segment
dTdt(2:end-1) = 1/mcp.*( UA_wall.*(u(3) - x(2:end-1)) + Qf_tes(1:end-1) ...
        - Qf_tes(2:end) + m_tank.*cp.*(x(1:end-2) - x(3:end))/2 );          % Mid segments

% Continuous system dynamics as a CasADi Function
f = Function('f',{x,u},{dTdt});

%% Define discrete system: x_next = F(x,u)
% Integrator options
intg_options = struct;
intg_options.number_of_finite_elements = n_c;
intg_options.simplify = true;
intg_options.tf = dt/intg_options.number_of_finite_elements;

% Reference Runge-Kutta implementation
intg = integrator('intg','rk',struct('x',x,'p',u,'ode',f(x,u)),intg_options);
res = intg('x0',x,'p',u);
xf = res.xf;

% Discretized (sampling time dt/n_c) system dynamics as a CasADi Function
F = Function('F', {x, u}, {xf});

```

**Comment by jgillis** [comment, marked as answer]
This failure occurs in a post-processing calculation of multipliers w.r.t. parameters.
Within `nlp_grad`, `gamma` is defined as $\gamma = \lambda_f f(x,p) + \lambda_g g(x,p) $.

Output `grad_gamma_p` computes $\nabla_p \gamma$.
The error indicates that this quantity is undefined for the chosen value of $p$.

Here is a minimal example that reproduces the error:
```python
from casadi import *

opti = Opti()

x = opti.variable()
p = opti.parameter()

opti.set_value(p, 0)
opti.minimize(x**2)

opti.subject_to(sin(x-sqrt(p))<=0)
opti.solver("ipopt")

sol = opti.solve()
opti.debug.casadi_solver.get_function('nlp_grad').disp(True)
```

A quick fix is to disable computation of `lam_p` by passing an option:
`opti.solver("ipopt",{"calc_lam_p":False})`


I've added this case to the FAQ here: https://github.com/casadi/casadi/wiki/FAQ:-Why-am-I-getting-%22NaN-detected%22in-my-optimization%3F
